Add Delta table support for `filesystem` destination

jorritsandbrink opened the pull request:
<!--
Thank you for submitting a pull request! Please provide a brief description of your changes below.
-->
### Description
This PR enables writing datasets to Delta tables in the `filesystem` destination.

A user can specify `delta` as `table_format` in a resource definition:

```py
@dlt.resource(table_name="a_delta_table", table_format="delta")
def a_resource():
    ...
```

<!--
Please link any related issues. This helps us keep the PR focused and merge it faster.
-->
### Related Issues

Contributes to #978

jorritsandbrink commented on dlt/common/destination/reference.py:
```diff
@@ -309,8 +309,12 @@ def restore_file_load(self, file_path: str) -> LoadJob:
         """Finds and restores already started loading job identified by `file_path` if destination supports it."""
         pass
 
+    def can_do_logical_replace(self, table: TTableSchema) -> bool:
```
Perhaps this can become a destination capability if we turn Delta into a full destination.

jorritsandbrink commented on dlt/destinations/impl/filesystem/filesystem.py:
```diff
@@ -41,56 +42,67 @@
+                client.fs_client.makedirs(self.pathlib.dirname(remote_path), exist_ok=True)
+            client.fs_client.put_file(local_path, remote_path)
 
     def make_remote_path(self) -> str:
```
A separation is made because Delta tables are managed at the folder level, while file formats are managed at the file level.

jorritsandbrink commented on dlt/pipeline/pipeline.py:
```diff
@@ -638,11 +638,40 @@ def run(
             LoadInfo: Information on loaded data including the list of package ids and failed job statuses. Please not that `dlt` will not raise if a single job terminally fails. Such information is provided via LoadInfo.
         """
+
+        def ensure_loader_file_format(
```
The current setup uses Parquet "staging" files that are written to a Delta table. If a user sets `table_format` to `delta`, then we set `loader_file_format` to `parquet`, overriding the value configured by the user (if any).

rudolfix replied:
please check my overall review. if we go for delta as a destination we could still keep this as a file format and this function could: (if file format is delta)
1. check if destination is `delta` and exit
2. check if destination is `filesystem` and swap it for `delta`
maybe let's write first version of the code and then we'll decide.

jorritsandbrink commented on tests/cases.py:
```diff
@@ -257,6 +257,11 @@ def assert_all_data_types_row(
             else:
                 db_mapping[binary_col] = bytes(db_mapping[binary_col])
 
+    # `delta` table format stores `wei` type as string
```
Decimals with precision higher than 38 are casted to string. `pyarrow` does not allow downcasting a decimal to a lower precision.

jorritsandbrink commented on dlt/destinations/impl/filesystem/filesystem.py:
```diff
@@ -455,3 +461,97 @@ def get_stored_schema(self) -> Optional[StorageSchemaInfo]:
+
+            assert isinstance(self.config.credentials, AwsCredentials)
+            storage_options = self.config.credentials.to_session_credentials()
+            storage_options["AWS_REGION"] = self.config.credentials.region_name
```
The `deltalake` library requires that `AWS_REGION` is provided. We need to add it to `DLT_SECRETS_TOML` under `[destination.filesystem.credentials]` to make `s3` tests pass on CI.

jorritsandbrink commented on dlt/destinations/impl/filesystem/filesystem.py:
```diff
@@ -455,3 +461,97 @@ def get_stored_schema(self) -> Optional[StorageSchemaInfo]:
+
+            assert isinstance(self.config.credentials, GcpServiceAccountCredentials)
+            gcs_creds = self.config.credentials.to_gcs_credentials()
+            gcs_creds["token"]["private_key_id"] = "921837921798379812"
```
This must be changed so that `private_key_id` is fetched from configuration.

jorritsandbrink commented on dlt/destinations/impl/filesystem/filesystem.py:
```diff
@@ -455,3 +461,97 @@ def get_stored_schema(self) -> Optional[StorageSchemaInfo]:
+            storage_options = self.config.credentials.to_session_credentials()
+            storage_options["AWS_REGION"] = self.config.credentials.region_name
+            # https://delta-io.github.io/delta-rs/usage/writing/writing-to-s3-with-locking-provider/#enable-unsafe-writes-in-s3-opt-in
+            storage_options["AWS_S3_ALLOW_UNSAFE_RENAME"] = "true"
```
Setting `AWS_S3_ALLOW_UNSAFE_RENAME` to `true` is the simplest setup. Perhaps we can later extend and let the user configure a locking provider. 

Context: https://delta-io.github.io/delta-rs/usage/writing/writing-to-s3-with-locking-provider/.

rudolfix replied:
interesting. we have a locking provider but probably not compatible with delta. it is called `transactional_file.py`

jorritsandbrink commented on dlt/common/destination/reference.py:
```diff
@@ -214,6 +214,20 @@ def exception(self) -> str:
         pass
 
 
+class DirectoryLoadJob:
```
Very minimal for now. Want to get some feedback before further polishing.

jorritsandbrink commented on dlt/common/storages/load_package.py:
```diff
@@ -177,6 +178,15 @@ def __str__(self) -> str:
         return self.job_id()
 
 
+class ParsedLoadJobDirectoryName(NamedTuple):
```
Also very minimal. Same as above.

jorritsandbrink commented on dlt/destinations/impl/filesystem/filesystem.py:
```diff
@@ -271,17 +303,25 @@ def start_file_load(self, table: TTableSchema, file_path: str, load_id: str) ->
     def restore_file_load(self, file_path: str) -> LoadJob:
         return EmptyLoadJob.from_file_path(file_path, "completed")
 
+    def start_dir_load(self, table: TTableSchema, dir_path: str, load_id: str) -> DirectoryLoadJob:
```
Perhaps `iceberg` will also be a directory job if we add it.

jorritsandbrink commented on dlt/destinations/impl/filesystem/filesystem.py:
```diff
@@ -455,3 +461,97 @@ def get_stored_schema(self) -> Optional[StorageSchemaInfo]:
+    def _write_delta_table(
+        self, path: str, table: "pa.Table", write_disposition: TWriteDisposition  # type: ignore[name-defined] # noqa
+    ) -> None:
+        """Writes in-memory Arrow table to on-disk Delta table."""
```
I introduced the concept of a "directory job", 1 is possible now. 2 is also possible. Both are tested in `test_pipeline_delta_filesystem_destination`. 3 seems not possible as discussed on the chat:
![image](https://github.com/dlt-hub/dlt/assets/47451109/4399d20d-c49c-4f3f-9eb3-2e15f54330d2)

jorritsandbrink commented on dlt/destinations/impl/filesystem/filesystem.py:
```diff
@@ -455,3 +461,97 @@ def get_stored_schema(self) -> Optional[StorageSchemaInfo]:
+        import pyarrow as pa
+        from deltalake import write_deltalake
+
+        def adjust_arrow_schema(
```
I moved them to `utils.py` and made them independent. Haven't added unit tests yet.

jorritsandbrink commented:
@rudolfix Can you review once more?

I addressed some of your feedback. Biggest changes since last review:
1. Introduction of the "directory job" so multiple files can be loaded at once. Thus far, the concept of a job was tightly coupled with a single file (if I understand correctly). But if there are multiple Parquet files, we want to be able to load them into the Delta table in a single commit. That is now possible. Code is still rogue/minimal, but I want some feedback before polishing it up.
2. Dedicated job for Delta table loads: `DeltaLoadFilesystemJob`.
3. Turned Delta-related methods into static utility methods.

jorritsandbrink commented on dlt/extract/storage.py:
```diff
@@ -24,11 +24,23 @@ def __init__(self, package_storage: PackageStorage, writer_spec: FileWriterSpec)
+        subfolder = self._get_data_item_subfolder(load_id, table_name)
         file_path = self.package_storage.get_job_file_path(
-            load_id, PackageStorage.NEW_JOBS_FOLDER, file_name
+            load_id, PackageStorage.NEW_JOBS_FOLDER, file_name, subfolder
```
This is where a folder gets included in the file path for directory jobs.

rudolfix reviewed (changes requested):
overall direction looks good! still several tasks are left but we have a clear direction now.

the big unified arrow table that is composed of many files in the dataset worries me a little. but we can do batching in a followup ticket.

jorritsandbrink commented on dlt/destinations/job_impl.py:
```diff
@@ -41,10 +41,10 @@ class EmptyLoadJob(EmptyLoadJobWithoutFollowup, FollowupJob):
         temp_file = os.path.join(tempfile.gettempdir(), self._file_name)
         with open(temp_file, "w", encoding="utf-8") as f:
-            f.write(data)
+            f.write("" if data is None else data)
```
Changed it. No empty string anymore.

jorritsandbrink commented on dlt/common/schema/typing.py:
```diff
@@ -64,7 +64,7 @@
 ]
 """Known hints of a column used to declare hint regexes."""
-TTableFormat = Literal["iceberg", "parquet", "jsonl"]
+TTableFormat = Literal["iceberg", "parquet", "jsonl", "delta"]
```
Done.

jorritsandbrink commented on dlt/load/load.py:
```diff
@@ -256,8 +256,13 @@ def create_followup_jobs(
                 if table_chain := get_completed_table_chain(
                     schema, all_jobs, top_job_table, starting_job.job_file_info().job_id()
                 ):
+                    table_jobs = [
```
Changed it. Child tables are handled now.

jorritsandbrink commented on dlt/destinations/impl/filesystem/utils.py:
```diff
@@ -0,0 +1,87 @@
+        from dlt.common.configuration.specs import GcpServiceAccountCredentials
+
+        assert isinstance(client.config.credentials, GcpServiceAccountCredentials)
+        gcs_creds = client.config.credentials.to_gcs_credentials()
```
Yes, we need to add `filesystem.credentials.private_key_id`. The test `test_gcp_object_store_rs_credentials` expects it.

jorritsandbrink commented on dlt/destinations/impl/filesystem/filesystem.py:
```diff
@@ -455,3 +494,14 @@ def get_stored_schema(self) -> Optional[StorageSchemaInfo]:
+        table_jobs: Optional[Sequence[LoadJobInfo]] = None,
+    ) -> List[NewLoadJob]:
+        jobs = super().create_table_chain_completed_followup_jobs(table_chain, table_jobs)
+        table_format = table_chain[0]["table_format"]
```
Fixed.

jorritsandbrink commented on dlt/destinations/impl/filesystem/filesystem.py:
```diff
@@ -41,70 +42,105 @@
+            _deltalake_storage_options,
+        )
+
+        file_paths = [job.file_path for job in self.table_jobs]
```
I don't think so. I don't think it's possible have non-Parquet files in the list.

jorritsandbrink commented on dlt/destinations/impl/filesystem/filesystem.py:
```diff
@@ -41,70 +42,105 @@
+            path=self.client.make_remote_uri(self.make_remote_path()),
+            data=ds.dataset(file_paths),
+            write_disposition=self.table["write_disposition"],
+            storage_options=_deltalake_storage_options(self.client),
```
I added `deltalake_storage_options` to `FilesystemConfiguration`.

jorritsandbrink commented on dlt/destinations/impl/filesystem/filesystem.py:
```diff
@@ -41,70 +42,105 @@
+
+        write_delta_table(
+            path=self.client.make_remote_uri(self.make_remote_path()),
+            data=ds.dataset(file_paths),
```
Added child table handling and a corresponding test.

jorritsandbrink commented on dlt/pipeline/pipeline.py:
```diff
@@ -638,11 +638,40 @@ def run(
             LoadInfo: Information on loaded data including the list of package ids and failed job statuses. Please not that `dlt` will not raise if a single job terminally fails. Such information is provided via LoadInfo.
         """
+
+        def ensure_loader_file_format(
```
I added two destination capabilities: `supported_table_formats` and `loader_file_format_adapter`. The latter can be used to adapt `preferred_loader_file_format` and `supported_loader_file_formats` at runtime.

jorritsandbrink commented:
@rudolfix I addressed your feedback. Can you review?

I changed the credentials handling for `deltalake` a bit since last time. It's a more generic setup now.

rudolfix commented on tests/load/pipeline/test_filesystem_pipeline.py:
```diff
@@ -216,6 +216,175 @@ def some_source():
 
 
+@pytest.mark.parametrize("multiple_files", (False, True))
+def test_pipeline_delta_filesystem_destination(
```
```python
@dlt.resource(primary_key="id", table_name=lambda i: i["type"])
    def _get_shuffled_events(
        last_created_at=dlt.sources.incremental("$", last_value_func=by_event_type)
    ):
        with open(
            "tests/normalize/cases/github.events.load_page_1_duck.json", "r", encoding="utf-8"
        ) as f:
            yield json.load(f)
```

rudolfix replied:
ofc remove incremental, you can yield json many times to create larger sets

jorritsandbrink replied:
Added a test called `test_delta_table_dynamic_dispatch` that does this.

rudolfix reviewed (changes requested):
getting there!
1. selecting file format should happen on the table level
2. we need a few more tests

rudolfix commented on dlt/normalize/normalize.py:
```diff
@@ -211,7 +248,8 @@ def _gather_metrics_and_close(
                     normalizer = _get_items_normalizer(
-                        DataWriter.item_format_from_file_extension(parsed_file_name.file_format)
+                        DataWriter.item_format_from_file_extension(parsed_file_name.file_format),
+                        stored_schema["tables"].get(root_table_name),
```
you do not handle `table_schema` being `None` in `_get_items_normalizer` - and this is a normal case (and we have failing low level tests IMO due to this)

jorritsandbrink replied:
Added handling for `table_schema is None`.

rudolfix commented on dlt/normalize/normalize.py:
```diff
@@ -110,33 +110,70 @@ def w_normalize_files(
+            def _get_items_normalizer(
+                item_format: TDataItemFormat, table_schema: TTableSchema
+            ) -> ItemsNormalizer:
+                if (item_format, table_schema["name"]) in item_normalizers:
```
in essence: now you create a normalizer per table name. so there's no reason to keep `item_format` in the tuple. I can't see any case where we write one part of the same table in one format and next in another.

so just cache by table name.

jorritsandbrink replied:
Changed it.

rudolfix commented on dlt/normalize/normalize.py:
```diff
@@ -110,33 +110,70 @@ def w_normalize_files(
+                    items_preferred_file_format, items_supported_file_formats = (
+                        destination_caps.loader_file_format_adapter(
+                            preferred_file_format,
+                            supported_file_formats,
```
maybe we should copy this list here... just in case someone modifies input list in place

rudolfix commented on dlt/normalize/normalize.py:
```diff
@@ -110,33 +110,70 @@ def w_normalize_files(
-            load_storage = LoadStorage(False, supported_formats, loader_storage_config)
+            load_storage = LoadStorage(False, supported_file_formats, loader_storage_config)
+
+            def _get_items_normalizer(
```
a side note: any more additions to this function and we'll extract it to utils to be unit tested.

jorritsandbrink commented on docs/website/docs/dlt-ecosystem/table-formats/delta.md:
```diff
@@ -0,0 +1,13 @@
+---
```
Do these table formats pages need to be included in a menu somewhere?

jorritsandbrink commented on dlt/normalize/normalize.py:
```diff
@@ -117,26 +117,50 @@ def w_normalize_files(
+        supported_file_formats = destination_caps.supported_loader_file_formats or []
+        supported_table_formats = destination_caps.supported_table_formats or []
+
+        for table in stored_schema["tables"].values():
```
Changed it.

jorritsandbrink commented on dlt/destinations/impl/filesystem/utils.py:
```diff
@@ -0,0 +1,79 @@
+    storage_options: Optional[Dict[str, str]] = None,
+) -> None:
+    """Writes in-memory Arrow table to on-disk Delta table."""
+    from deltalake import write_deltalake
```
Moved it to `dlt/common/libs/deltalake.py`.

jorritsandbrink commented on dlt/destinations/impl/filesystem/utils.py:
```diff
@@ -0,0 +1,79 @@
+from typing import Optional, Dict, Union
+
+import pyarrow as pa
```
Done.

jorritsandbrink commented on dlt/common/libs/pyarrow.py:
```diff
@@ -397,6 +400,29 @@ def pq_stream_with_new_columns(
+    return data.to_table() if isinstance(data, pyarrow.dataset.Dataset) else data
+
+
+def adjust_arrow_schema(
```
Changed it.

jorritsandbrink commented on dlt/common/configuration/specs/aws_credentials.py:
```diff
@@ -45,6 +45,14 @@ def to_session_credentials(self) -> Dict[str, str]:
+        # https://docs.rs/object_store/latest/object_store/aws
+        assert self.region_name is not None, "`object_store` Rust crate requires AWS region."
+        creds = self.to_session_credentials()
+        if creds["aws_session_token"] is None:
```
Correct.

rudolfix reviewed (dismissed):
LGTM! amazing work!

I added a few small fixes mostly to handle failing pydantic tests. see commits for details

rudolfix commented on tests/load/pipeline/test_filesystem_pipeline.py:
```diff
@@ -216,6 +222,244 @@ def some_source():
         assert table.column("value").to_pylist() == [1, 2, 3, 4, 5]
 
 
+@pytest.mark.essential
```
1. if we add a new feature, it makes sense to write 1-2 tests with happy paths and make them as essential
2. if tests are finishing quickly (like all the local pipeline tests here, they also can be marked as essential)

essential is used to have a fast smoke-testing for destinations in normal CI runs
